src/wasm-traversal.h

```diff
@@ -204,12 +205,14 @@ struct Walker : public VisitorType {
     static_cast<SubType*>(this)->visitTable(table);
   }
 
-  void walkMemory(Memory* memory) {
-    for (auto& segment : memory->segments) {
-      if (!segment.isPassive) {
-        walk(segment.offset);
-      }
+  void walkDataSegment(DataSegment* segment) {
+    if (!segment->isPassive) {
+      walk(segment->offset);
     }
+    static_cast<SubType*>(this)->visitDataSegment(segment);
+  }
+
+  void walkMemory(Memory* memory) {
     static_cast<SubType*>(this)->visitMemory(memory);
```

Comment:
This matches `walkTable`, but reading the old code now, it looks wrong. `walk` should visit all children, not just `visit` the parent and stop. We should fix that eventually, I guess. For now maybe let's add a TODO so we don't forget,
```suggestion
    // TODO: This method and walkTable should walk children too, or be renamed.
    static_cast<SubType*>(this)->visitMemory(memory);
```